### PR TITLE
[WIP]  Added subgrid variability in P3 through SHOC

### PR DESCRIPTION
**Pull request by JS-WRF-SBM**
- Subgrid variability is introduced into P3 through SHOC 
- Tuning parameters for the relative variance per process are included and defined in the namelist

### CODE REVIEW

**Comment by JS-WRF-SBM** [comment]
@AaronDonahue @PeterCaldwell @kaizhangpnl 

Hi All,
Please review this PR and make ANY suggestion -- will follow up for this PR and especially for the next PRs. 

Thanks,
Kobby

File: components/cam/src/physics/cam/shoc.F90 [hunk 1]
```diff
@@ -1631,6 +1636,12 @@ subroutine shoc_assumed_pdf(&
       ! Compute SGS liquid water mixing ratio
       shoc_ql(i,k) = max(0._rtype,a*ql1+(1._rtype-a)*ql2)
 
+! +++ JShpund: Add cloud liquid variance (CLUBB formulation, adjusted to SHOC parameters based on Peter B.)
+!              * Please double check this *
+      shoc_ql2(i,k) = a * ( s1*ql1 + C1*std_s1**2.0 )                  &
+                    + ( 1._rtype-a ) * ( s2*ql2 + C2*std_s2**2.0 ) - shoc_ql(i,k)**2.0
+      shoc_ql2(i,k) = max( 0._rtype, shoc_ql2(i,k) )
+
```
Comment (PeterCaldwell) [inline components/cam/src/physics/cam/shoc.F90:1644]: @bogensch - please confirm that the shoc-related portion of this PR is correct.

**Comment by PeterCaldwell** [comment]
Looks pretty good, @JS-WRF-SBM ! Thanks. I haven't had time to really think carefully about it yet, but a few initial impressions:

1. The point of this PR is that if qc follows a gamma distribution, the in-cloud average of a local process rate qc^b is gamma(relvar + b)/gamma(relvar)*relvar^b times the in-cloud average qc^b. You have added the various exponents "b" for the relevant processes to the namelist and are using those coefficients for getting the subgrid-variability scaling factor for each process (which you call local_coef)... but you're still using the hard-coded coefficients for the actual qc^b calculation! This will give wrong answers unless you happen to set the parameters equal to their hard-coded values...

2. Your names for these exponents for each process are wacky. I suggested "p3_prc_<stuff>" as a template for autoconversion because "prc" is MG2's name for autoconversion. Don't use it for other processes. In fact, better names would be something like: p3_CldImmFrz_Expon, p3_QcAccret_Expon, p3_QrAccret_Expon, and p3_QcAutoCon_Expon.

3. In general, I'd encourage you to try to use more consistent/understandable names. For example, can you call relvar => qc_relvar, local_coef=>subgrid_variab_coef (or something better - that's not great). I also see that what SHOC calls ql is called qc by P3... I'll talk to Peter et al about whether we want to make this consistent.

**Comment by PeterCaldwell** [comment]
Also, it looks like there are merge conflicts in your code (if you search on this page, you will see "This branch has conflicts that must be resolved". It looks like they're related to @AaronDonahue 's recent PR that deleted stuff in the P3 interface that wasn't used. @JS-WRF-SBM : please check whether you need these variables. I think you don't except for relvar.

**Comment by JS-WRF-SBM** [comment]
> Looks pretty good, @JS-WRF-SBM ! Thanks. I haven't had time to really think carefully about it yet, but a few initial impressions:
> 
>     1. The point of this PR is that if qc follows a gamma distribution, the in-cloud average of a local process rate qc^b is gamma(relvar + b)/gamma(relvar)*relvar^b times the in-cloud average qc^b. You have added the various exponents "b" for the relevant processes to the namelist and are using those coefficients for getting the subgrid-variability scaling factor for each process (which you call local_coef)... but you're still using the hard-coded coefficients for the actual qc^b calculation! This will give wrong answers unless you happen to set the parameters equal to their hard-coded values...

@PeterCaldwell 
That's correct, but only temporary as I was asked to submit the PR as "stand alone".
You are referring to the Accretion and the Auto-conversion processes, for which a different PR is pending with these hard-coded values as tuning parameters. 
 

> 
>     2. Your names for these exponents for each process are wacky. I suggested "p3_prc_" as a template for autoconversion because "prc" is MG2's name for autoconversion. Don't use it for other processes. In fact, better names would be something like: p3_CldImmFrz_Expon, p3_QcAccret_Expon, p3_QrAccret_Expon, and p3_QcAutoCon_Expon.

OK. 

> 
>     3. In general, I'd encourage you to try to use more consistent/understandable names. For example, can you call relvar => qc_relvar, local_coef=>subgrid_variab_coef (or something better - that's not great). I also see that what SHOC calls ql is called qc by P3... I'll talk to Peter et al about whether we want to make this consistent.

OK. Let me know what is the final preference here.

**Comment by JS-WRF-SBM** [comment]
> Also, it looks like there are merge conflicts in your code (if you search on this page, you will see "This branch has conflicts that must be resolved". It looks like they're related to @AaronDonahue 's recent PR that deleted stuff in the P3 interface that wasn't used. @JS-WRF-SBM : please check whether you need these variables. I think you don't except for relvar.

@PeterCaldwell @AaronDonahue 
Aaron, have possibly erased some 'relvar' / 'relvar_idx' related settings in the 'micro_p3_interface.f90'?
 
Before your PR (from top to bottom):
```
relvar_idx
call pbuf_add_field('RELVAR',     'global',dtype_r8,(/pcols,pver/), relvar_idx)
call pbuf_set_field(pbuf2d, relvar_idx, 2._rtype)
real(rtype), pointer :: relvar(:,:)       ! relative variance of cloud water
call pbuf_get_field(pbuf, relvar_idx,      relvar,      col_type=col_type, copy_if_needed=use_subcol_microp)
```
After your PR:
```
relvar_idx
call pbuf_add_field('RELVAR',     'global',dtype_r8,(/pcols,pver/), relvar_idx)
call pbuf_set_field(pbuf2d, relvar_idx, 2._rtype)  
```

**Comment by AaronDonahue** [comment]
> > Also, it looks like there are merge conflicts in your code (if you search on this page, you will see "This branch has conflicts that must be resolved". It looks like they're related to @AaronDonahue 's recent PR that deleted stuff in the P3 interface that wasn't used. @JS-WRF-SBM : please check whether you need these variables. I think you don't except for relvar.
> 
> @PeterCaldwell @AaronDonahue
> Aaron, have possibly erased some 'relvar' / 'relvar_idx' related settings in the 'micro_p3_interface.f90'?
> 
> Before your PR (from top to bottom):
> 
> ```
> relvar_idx
> call pbuf_add_field('RELVAR',     'global',dtype_r8,(/pcols,pver/), relvar_idx)
> call pbuf_set_field(pbuf2d, relvar_idx, 2._rtype)
> real(rtype), pointer :: relvar(:,:)       ! relative variance of cloud water
> call pbuf_get_field(pbuf, relvar_idx,      relvar,      col_type=col_type, copy_if_needed=use_subcol_microp)
> ```
> 
> After your PR:
> 
> ```
> relvar_idx
> call pbuf_add_field('RELVAR',     'global',dtype_r8,(/pcols,pver/), relvar_idx)
> call pbuf_set_field(pbuf2d, relvar_idx, 2._rtype)  
> ```

@JS-WRF-SBM, yes.  Until now, relvar was not used in p3_interface and my PR was essentially a big cleanup.  You should be able to add the two lines:
```
real(rtype), pointer :: relvar(:,:)       ! relative variance of cloud water
call pbuf_get_field(pbuf, relvar_idx,      relvar,      col_type=col_type, copy_if_needed=use_subcol_microp)
```
back into p3_micro_interface.F90.  You might need to rebase your PR, at which point these two lines will be deleted.  And then go back in and add them.  I can help you with this if you would like since I did the original cleanup and may have a better sense of what is missing that you need back.

**Comment by PeterCaldwell** [comment]
About having different constants for the exponent versus for the SGS variability coefficient: Ok, I see why you saved the exponent change for later. However, replacing the hard-coded exponents with the values you're now getting from the namelist is just a trivial 1 line change in each place and making that change keeps the code in a consistent/scientifically credible state. In retrospect, if you wanted to make 2 PRs it would have made sense to make the namelist exponent change first and then add the SGS variability. But I think 1 PR for both changes is fine.

Does what I said about replacing the exponents with your coefficients make sense? For example, for autoconversion I'm thinking you would replace 2.47 with p3_prc_relvar_AutoCon (or whatever you rename that parameter to). Though looking at what you've done, I see that you're setting the default for p3_prc_relvar_AutoCon to 3.19, which I guess is Po-Lun's new tuning for it. Could you instead use the old tunings (e.g. 2.47 for autoconversion) as the defaults in this PR, then make a new mini-PR to use different defaults for SCREAM-LR and FC5AV1C compsets?

**Comment by JS-WRF-SBM** [comment]
> About having different constants for the exponent versus for the SGS variability coefficient: Ok, I see why you saved the exponent change for later. However, replacing the hard-coded exponents with the values you're now getting from the namelist is just a trivial 1 line change in each place and making that change keeps the code in a consistent/scientifically credible state. In retrospect, if you wanted to make 2 PRs it would have made sense to make the namelist exponent change first and then add the SGS variability. But I think 1 PR for both changes is fine.

@PeterCaldwell  
OK, will do it in 1 PR.
 
> Does what I said about replacing the exponents with your coefficients make sense? For example, for autoconversion I'm thinking you would replace 2.47 with p3_prc_relvar_AutoCon (or whatever you rename that parameter to). Though looking at what you've done, I see that you're setting the default for p3_prc_relvar_AutoCon to 3.19, which I guess is Po-Lun's new tuning for it. Could you instead use the old tunings (e.g. 2.47 for autoconversion) as the defaults in this PR, then make a new mini-PR to use different defaults for SCREAM-LR and FC5AV1C compsets?

Here, lets make sure we are on the same page. I'll leave the original tuning values intact for this PR.
For autoconversion, the rate of liquid mass changes according to:
`qcaut = 1350._rtype*bfb_pow(dum,2.47_rtype)*bfb_pow(nc_incld*1.e-6_rtype*rho,-1.79_rtype)`
After my change (leaving aside the naming convention for now):
`qcaut = Fact_relvar*Fact_1*bfb_pow(dum,Expo_1)*bfb_pow(nc_incld*1.e-6_rtype*rho,Expo_2)`
Is that what you had in mind?

**Comment by PeterCaldwell** [comment]
@JS-WRF-SBM : sorry I wasn't clear in my previous comment by saying in my 1st paragraph that you should have 1 PR and in the 2nd that you should have 2. I think you should have 2 PRs. 

In the PR you're working on now, for autoconversion, you currently have:
`local_coef = var_coef(ql_relvar, p3_prc_relvar_AutoCon)`
 `qcaut = local_coef*1350._rtype*bfb_pow(dum,2.47_rtype)*bfb_pow(nc_incld*1.e6_rtype*rho,-1.79_rtype)`

The only change I'm suggesting you make is to change 2.47_rtype in that 2nd line to p3_prc_relvar_AutoCon
`qcaut = local_coef*1350._rtype*bfb_pow(dum,p3_prc_relvar_AutoCon)*bfb_pow(nc_incld*1.e6_rtype*rho,-1.79_rtype)`

Then, in a second PR, to also convert the 1350 and -1.79 to namelist parameters and to set their default values appropriately for different compsets. My motivation for doing 2 PRs is that getting namelist parameters to work right for different compsets can be fiddly work and breaking down steps into the smallest chunks possible minimizes the chances of getting confused/stuck tracking down a problem.

**Comment by JS-WRF-SBM** [comment]
> @JS-WRF-SBM : sorry I wasn't clear in my previous comment by saying in my 1st paragraph that you should have 1 PR and in the 2nd that you should have 2. I think you should have 2 PRs.
> 
> In the PR you're working on now, for autoconversion, you currently have:
> `local_coef = var_coef(ql_relvar, p3_prc_relvar_AutoCon)`
> `qcaut = local_coef*1350._rtype*bfb_pow(dum,2.47_rtype)*bfb_pow(nc_incld*1.e6_rtype*rho,-1.79_rtype)`
> 
> The only change I'm suggesting you make is to change 2.47_rtype in that 2nd line to p3_prc_relvar_AutoCon
> `qcaut = local_coef*1350._rtype*bfb_pow(dum,p3_prc_relvar_AutoCon)*bfb_pow(nc_incld*1.e6_rtype*rho,-1.79_rtype)`
> 
> Then, in a second PR, to also convert the 1350 and -1.79 to namelist parameters and to set their default values appropriately for different compsets. My motivation for doing 2 PRs is that getting namelist parameters to work right for different compsets can be fiddly work and breaking down steps into the smallest chunks possible minimizes the chances of getting confused/stuck tracking down a problem.

@PeterCaldwell 
OK, will do for autoconversion.
2 followup questions:
1. There are 2 more processes that involves the subgrid 'relvar' change being factored a power law constants which can be tuned, particularity accretion. Should we perform the above only for autoconversion?
2. Would like the 2nd PR to be checkout from the 'master' or the 1st PR?

**Comment by PeterCaldwell** [comment]
1. I was using autoconversion as an example - you should do the same thing to immersion freezing and accretion. Again, just use the parameters you've created for the SGS variability addition for the exponents and don't bother making namelist parameters of the other terms (yet).
2. About whether to create your next PR from master or from this branch: I generally try to create PRs from master. I guess there wouldn't be a problem if you create a branch off this branch and this branch is merged to master before your new branch is (though you'd probably need to rebase the new branch to grab other work people have done in the meantime). I'll ask on slack.

**Pull request closed without merging**